### PR TITLE
Support for breakpoint debugging

### PR DESCRIPTION
**Pull request by ohodson**

### CODE REVIEW

File: src/workerd/io/worker.c++ [hunk 1]
```diff
@@ -2256,8 +2146,220 @@ public:
...
+    }
+
+    kj::Promise<void> incomingLoop() {
+      return webSocket.receive().then([this](kj::WebSocket::Message&& message) -> kj::Promise<void> {
+        KJ_SWITCH_ONEOF(message) {
+          KJ_CASE_ONEOF(text, kj::String) {
+            auto lockedIncomingMessages = incomingQueue.lockExclusive();
+            lockedIncomingMessages->messages.add(kj::mv(text));
+            if (lockedIncomingMessages->messages.size() - lockedIncomingMessages->head == 1) {
+              isolateExecutor.executeAsync([this]() {
```
[the remark this answers is not in the document]
Comment (ohodson) [reply src/workerd/io/worker.c++:2335]: Yep, can't depend on an executor captured earlier being valid later.

**Comment by ohodson** [comment]
PTAL, this is a moderately large change that we'll likely have to iterate on quite a bit.

The major shortcoming here is that this is only worker for single service configs, with multiple services we attempt to launch multiple inspector threads on the same port. This should be fixable, but it would be good to feedback in the interim.

Thanks.

**Comment by ohodson** [comment]
> The major shortcoming here is that this is only worker for single service configs. This should be fixable, but it would be good to get feedback in the interim. The specific example that fails is ssh://[git@bitbucket.cfdata.org](mailto:git@bitbucket.cfdata.org):7999/~bcoll/workerd-bus-error-10.git

This is addressed in the second commit. Wrangler does not expect to package multiple user written workers in a config, but may in practice. With the second commit chrome devtools would be able to debug any code referenced by the config. Wrangler devtools only wants to be able to debug the user written workers (except if they need to debug their stuff).

File: src/workerd/io/worker.c++ [hunk 2]
```diff
@@ -448,37 +448,106 @@ public:
...
-      if (info->threadId == getCurrentThreadId()) {
-        // We're on an inspector-serving thread.
+    } else {
+      auto lockedState = state.lockExclusive();
+      KJ_IF_MAYBE(info, lockedState->inspectorTimerInfo) {
         timePoint = info->timer.now() + info->timerOffset
                   - kj::origin<kj::TimePoint>() + kj::UNIX_EPOCH;
+      } else {
+        // We're at script startup time -- just return the Epoch.
       }
```
Comment (jasnell) [inline src/workerd/io/worker.c++:458]: non-blocking nit:

```suggestion
      }
      // We're at script startup time -- just return the Epoch.
```

File: src/workerd/io/worker.c++ [hunk 3]
```diff
@@ -448,37 +448,106 @@ public:
...
-  void setInspectorTimerInfo(kj::Timer& timer, kj::Duration timerOffset) {
-    // Helper for attachInspector().
-    inspectorTimerInfo = InspectorTimerInfo { timer, timerOffset, getCurrentThreadId() };
+  void runMessageLoopOnPause(int contextGroupId) override {
+    // This method is called by v8 when a breakpoint or debugger statement is hit. This method
+    // processes debugger messages until `Debugger.resume()` is called, when v8 then calls
+    // `quitMessageLoopOnPause()`.
+    //
+    //NB This method is ultimately called from the `InspectorChannelImpl` and the isolate lock is
+    // held when this method is called.
```
Comment (jasnell) [inline src/workerd/io/worker.c++:484]: What's the `NB` mean here?

File: src/workerd/io/worker.c++ [hunk 4]
```diff
@@ -448,37 +448,106 @@ public:
...
+    if (isMultiTenantProcess()) {
+      // TODO: breakpoints are not supported for the multi-tenant case.
```
Comment (jasnell) [inline src/workerd/io/worker.c++:486]: Is this really a `TODO`?

**Review by jasnell — APPROVED** [review]

File: src/workerd/io/worker.c++ (hunk 2 above)
Comment (ohodson) [reply src/workerd/io/worker.c++:458]: Thanks, incorporated.

File: src/workerd/io/worker.c++ (hunk 3 above)
Comment (ohodson) [reply src/workerd/io/worker.c++:484]: Nota Bene, old man english, dropped.

File: src/workerd/io/worker.c++ (hunk 4 above)
Comment (ohodson) [reply src/workerd/io/worker.c++:486]: Yes, I'm afraid so. We continue to pass ew tests with this change, but I don't know enough about multi-tenant to know how well this would work at this point.

Specific gaps are:

1) For breakpoints you really need source maps to get started. That's easy to cons-up for local testing (add "//# sourceURL=file:///..."), but for ew it may require additional plumbing, e.g. fetch network resource.

2) What the expectation should be if with multiple requests and breakpoints / debugger statements. Naively, I assume we may end up with multiple isolates for the same scripts running on a single machine.

3) How to deploy a custom build of ew and test the ew preview case and test connecting CDT? https://wiki.cfdata.org/display/EW/How+to+debug+a+custom+edgeworker+build looks like a reasonable starting point.

File: src/workerd/server/server.c++ [hunk 5]
```diff
@@ -964,6 +964,27 @@ kj::Own<Server::Service> Server::makeDiskDirectoryService(
 
 // =======================================================================================
 
+class Server::InspectorServiceIsolateRegistrar final {
```
Comment (kentonv) [inline src/workerd/server/server.c++:967]: Could use a doc comment.

File: src/workerd/server/server.c++ [hunk 6]
```diff
@@ -964,6 +964,27 @@ kj::Own<Server::Service> Server::makeDiskDirectoryService(
...
+public:
+  InspectorServiceIsolateRegistrar() {}
+  ~InspectorServiceIsolateRegistrar();
```
Comment (kentonv) [inline src/workerd/server/server.c++:970]: KJ style says all destructors should be declared either `noexcept(false)`, or if noexcept behavior is explicitly desired, then `noexcept`.

(I hate that C++11 changed the default.)

**Review by kentonv — COMMENTED** [review]
Just started looking at this, realize I need more time than I have now, so posting the couple comments I already wrote but I'll try to look at the whole thing either later today or tomorrow.

File: src/workerd/io/worker.c++ [hunk 7]
```diff
@@ -2420,75 +2536,183 @@ private:
...
+
+  // We can be paused by either hitting a debugger statement in a script or from hitting
+  // a breakpoint or someone hit break. As a result we may or may not be recursive dispatching messages.
+
+  KJ_IF_MAYBE(message, maybeMessage) {
+    auto lockedState = this->state.lockExclusive();
+    if (IoContext::hasCurrent()) {
+      // Received a message whilst script is running, probably in a breakpoint.
+      v8_inspector::V8InspectorSession& session = *lockedState->get()->session;
+      // const_cast OK because we're going to lock it
```
Comment (kentonv) [inline src/workerd/io/worker.c++:2690]: I guess more accurately, const_cast OK because we already have the lock via the IoContext?

File: src/workerd/io/worker.c++ [hunk 8]
```diff
@@ -2326,8 +2214,236 @@ public:
...
+  // a breakpoint or debugger statement.
+
+  bool dispatchOneMessageDuringPause();
+  // Dispatches one message whilst automatic CDP messages on the I/O worker thread is paused, called
+  // on the thread executing the isolate whilst execution is suspended due to a breakpoint or
+  // debugger statement.
+
 private:
-  kj::WebSocket& webSocket;
+  class WebSocketIoWorker final {
```
Comment (kentonv) [inline src/workerd/io/worker.c++:2231]: The use of the word "Worker" in this name is a bit confusing given our product name. :)

When I saw references to "I/O worker thread" in comments above I initially assumed this meant "the thread running the worker".

Maybe "Handler" instead of "Worker"? Or some other generic word, just not "Worker".

File: src/workerd/io/worker.c++ [hunk 9]
```diff
@@ -2326,8 +2214,236 @@ public:
...
+    // the InspectorChannelImpl and the InspectorClient.
+  public:
+    WebSocketIoWorker(kj::WebSocket& webSocket)
+        : webSocket(webSocket), receivedClose(false) {
+      // Assume we are being instantiated on the InspectorService thread, the thread that will do
+      // I/O for CDP messages. Messages are delivered to the InspectorChannelImpl on the Isolate thread.
+      outgoingQueueNotifier = kj::atomicRefcounted<XThreadNotifier>(kj::getCurrentThreadExecutor());
+    }
+
+    ~WebSocketIoWorker() {
```
Comment (kentonv) [inline src/workerd/io/worker.c++:2244]: `noexecpt(false)`

File: src/workerd/io/worker.c++ [hunk 10]
```diff
@@ -448,37 +448,105 @@ public:
...
+    {
+      auto lockedState = state.lockExclusive();
+      KJ_IF_MAYBE(channel, lockedState->channel) {
+        pauseIncomingMessages(*channel);
+      }
+    }
+
+    runMessageLoop = true;
+    do {
+      auto lockedState = state.lockExclusive();
```
Comment (kentonv) [inline src/workerd/io/worker.c++:498]: It looks like this lock is held while we wait for the next message to arrive, and then is dropped only momentarily at the end of the loop and taken back a moment later.

If any other threads are actually contending for the lock this seems like a problem. Probably those other threads won't actually wake up fast enough to take the lock before the paused thread grabs it again. And if no subsequent message happens to be delivered, those threads could be waiting indefinitely.

If this isn't a problem could you expand the comments here to explain how so?

File: src/workerd/io/worker.c++ [hunk 11]
```diff
@@ -448,37 +448,105 @@ public:
...
+
   struct InspectorTimerInfo {
     kj::Timer& timer;
     kj::Duration timerOffset;
     uint64_t threadId;
   };
 
-  kj::Maybe<InspectorTimerInfo> inspectorTimerInfo;
-  // The timer and offset for the inspector-serving thread.
+  volatile bool runMessageLoop;
```
Comment (kentonv) [inline src/workerd/io/worker.c++:537]: Is this really volatile? It looks like it will only be accessed in methods that run under isolate lock.

**Review by kentonv — COMMENTED** [review]
(still reading, but posting a batch before going to more meetings...)

File: src/workerd/io/worker.c++ [hunk 12]
```diff
@@ -448,37 +448,105 @@ public:
...
-    // time -- just return the Epoch.
+  void setInspectorTimerInfo(kj::Timer& timer, kj::Duration timerOffset) {
+    auto lockedState = state.lockExclusive();
+    lockedState->inspectorTimerInfo = InspectorTimerInfo { timer, timerOffset, getCurrentThreadId() };
+  }
 
-    return (timePoint - kj::UNIX_EPOCH) / kj::MILLISECONDS;
+  void setChannel(Worker::Isolate::InspectorChannelImpl& channel) {
+    auto lockedState = state.lockExclusive();
+    lockedState->channel = channel;
```
Comment (kentonv) [inline src/workerd/io/worker.c++:469]: What if there's an existing connection and so `lockedState->channel` is non-null here? Looks like when the other connection later dies, it'll call `resetChannel()` which will incorrectly clear `lockedState->channel`?

File: src/workerd/io/worker.c++ [hunk 13]
```diff
@@ -2420,75 +2536,183 @@ private:
...
+      kj::String message) {
+    v8_inspector::V8InspectorSession& session = *lockedState->get()->session;
+    // const_cast OK because we're going to lock it
+    Isolate& isolate = const_cast<Isolate&>(*lockedState->get()->isolate);
+    jsg::V8StackScope stackScope;
+    Isolate::Impl::Lock recordedLock(isolate, InspectorLock(nullptr), stackScope);
+    dispatchProtocolMessage(kj::mv(message), session, isolate, stackScope, recordedLock);
+  }
+
+  void dispatchProtocolMessage(kj::String message,
```
Comment (kentonv) [inline src/workerd/io/worker.c++:2553]: Looks like the code in this method originates from `incomingLoop()`. Any chance you can move this method up to be in the same place in the class where `incomingLoop()` used to be, so that it's easier to see the diff? We could move it back down here (to be with other `private` stuff) in a separate commit (probably wait to do that until the code is ready to merge, or even do it in a separate PR).

File: src/workerd/io/worker.c++ (hunk 7 above)
Comment (ohodson) [reply src/workerd/io/worker.c++]: Thanks, done.

File: src/workerd/io/worker.c++ (hunk 9 above)
Comment (ohodson) [reply src/workerd/io/worker.c++]: Thanks, done.

File: src/workerd/io/worker.c++ (hunk 11 above)
Comment (ohodson) [reply src/workerd/io/worker.c++]: Oops, done, that's a very early debugging bandaid.

File: src/workerd/io/worker.c++ (hunk 8 above)
Comment (ohodson) [reply src/workerd/io/worker.c++]: Thanks, done.

File: src/workerd/server/server.c++ (hunk 5 above)
Comment (ohodson) [reply src/workerd/server/server.c++:967]: Thanks, done.

File: src/workerd/server/server.c++ [hunk 14]
```diff
@@ -972,12 +993,26 @@ class Server::InspectorService final: public kj::HttpService, public kj::HttpSer
 public:
   InspectorService(
       kj::Timer& timer,
-      kj::HttpHeaderTable::Builder& headerTableBuilder)
+      kj::HttpHeaderTable::Builder& headerTableBuilder,
+      InspectorServiceIsolateRegistrar& theRegistrar)
```
[the remark this answers is not in the document]
Comment (ohodson) [reply src/workerd/server/server.c++]: Okay, done

File: src/workerd/server/server.c++ (hunk 6 above)
Comment (ohodson) [reply src/workerd/server/server.c++]: Thanks, done

File: src/workerd/io/worker.c++ [hunk 15]
```diff
@@ -2420,75 +2536,183 @@ private:
...
+      Isolate::Impl::Lock& recordedLock = workerLock.impl->recordedLock;
+      jsg::V8StackScope stackScope;
+      dispatchProtocolMessage(kj::mv(*message), session, isolate, stackScope, recordedLock);
     } else {
-      auto first = kj::mv(messages[0]);
-      auto rest = messages.slice(1, messages.size());
-      return webSocket.send(first).attach(kj::mv(first)).then([this, rest]() mutable {
-        return sendToWebsocket(rest);
-      });
+      // Received a message whilst idle.
```
[the remark this answers is not in the document]
Comment (ohodson) [reply src/workerd/io/worker.c++:2692]: This code has been hacked around quite a bit, I suspect we don't need this at all at this point.

File: src/workerd/io/worker.c++ (hunk 12 above)
Comment (ohodson) [reply src/workerd/io/worker.c++:472]: Comment and invariant added. I might be misunderstanding, but I don't think this is a problem in practice
due to there only being one connection at a time (https://github.com/cloudflare/workerd/blob/7b97887d6dde83da067caaaac946eb6421d5a36c/src/workerd/io/worker.c%2B%2B#L2521).

File: src/workerd/io/worker.c++ (hunk 4 above)
Comment (ohodson) [reply src/workerd/io/worker.c++]: Dropped. I was really looking for a way to distinguish product from workerd since I would not consider this production ready. Probably better to just grok the process name.

File: src/workerd/io/worker.c++ (hunk 13 above)
Comment (ohodson) [reply src/workerd/io/worker.c++]: Done.

File: src/workerd/io/worker.c++ (hunk 15 above)
[the remark this answers is not in the document]
Comment (ohodson) [reply src/workerd/io/worker.c++]: Dropped.

File: src/workerd/io/worker.c++ (hunk 10 above)
Comment (kentonv) [reply src/workerd/io/worker.c++:510]: So from the comment it sounds like this only matters on disconnect/shutdown? But won't shutdown cause dispatchOneMessageDuringPause() to return false, causing the loop here to break entirely? I still don't see why we need to release the lock for each iteration of the loop, rather than just at the end of the loop.

On the other hand though, I really feel like as a rule we shouldn't be waiting for I/O while holding a mutex lock. Is there a way we could structure this code so that it doesn't hold the lock while waiting for messages?

Comment (ohodson) [reply src/workerd/io/worker.c++]: I think you are right about the shutdown path. The last push fixes that.

I may be able to see a path to avoid the mutex...right now it is preventing deletion of the channel whilst processing messages, but if it is more closely associated with the WebSocketIoHandler this might be avoided.

**Review by kentonv — COMMENTED** [review]
I had some review comments that seem to have disappeared, but I also notice that github says I have a pending review, but it says there are no comments, but maybe if I post it anyway they will come back? Trying...

**Comment by kentonv** [comment]
Sadly, it appears my comments were lost. Ugh.

File: src/workerd/io/worker.c++ [hunk 16]
```diff
@@ -2327,8 +2311,236 @@ public:
...
+    }
+
+    kj::Maybe<kj::String> waitForMessage() {
+      // Blocked the current thread until a message arrives. This is intended
+      // for use in the InspectorClient when breakpoints are hit. The InspectorClient
+      // has to remain in runMessageLoopOnPause() but still receive CDP messages
+      // (e.g. resume).
+      return incomingQueue.when(
+        [this](const MessageQueue& incomingQueue) {
+          return (incomingQueue.head < incomingQueue.messages.size() || isClosed());
```
Comment (kentonv) [inline src/workerd/io/worker.c++:2377]: The comments I tried to post twice before and github lost it both times, it seems:

It feels weird that this conditional wait is considering `isClosed()`, which isn't part of the mutex state. The condition won't automatically be checked when `isClosed()` changes state, it's only checked when the mutex is unlocked implying that something could have changed. I see that in practice you always lock and unlock this mutex shortly after changing the isClosed() flag, so it works out in practice. But, perhaps it would be less brittle to make the closed state be part of the `incomingQueue` mutex?

Note that if you need to access it atomically somewhere where locking the mutex would be too much overhead, you can use `getWithoutLock()` to skip the lock (and this should be safe enough when atomically reading a boolean).

Comment (ohodson) [reply src/workerd/io/worker.c++]: Thanks, added a status field to the queue definition (active, paused, closed).

**Comment by ohodson** [comment]
@kentonv, friendly ping, last comment incorporated on 7/14, thanks.

File: src/workerd/io/worker.c++ [hunk 17]
```diff
@@ -2317,8 +2301,243 @@ public:
...
+        lockedIncomingQueue->status = MessageQueue::Status::CLOSED;
+      }
+      {
+        auto lockedOutgoingQueue = outgoingQueue.lockExclusive();
+        lockedOutgoingQueue->status = MessageQueue::Status::CLOSED;
+      }
+    }
+
+    kj::Promise<void> incomingLoop() {
+      return webSocket.receive().then([this](kj::WebSocket::Message&& message) -> kj::Promise<void> {
```
Comment (kentonv) [inline src/workerd/io/worker.c++:2442]: Optional: Might be nice to convert to a coroutine with a for() loop? Also for `outgoingLoop()`.

File: src/workerd/io/worker.c++ [hunk 18]
```diff
@@ -449,37 +449,84 @@ public:
       // We're on a request-serving thread.
       auto& ioContext = IoContext::current();
       timePoint = ioContext.now();
-    } else KJ_IF_MAYBE(info, inspectorTimerInfo) {
-      if (info->threadId == getCurrentThreadId()) {
-        // We're on an inspector-serving thread.
+    } else {
+      auto lockedState = state.lockExclusive();
```
Comment (kentonv) [inline src/workerd/io/worker.c++:453]: Dues this ever get called recursively within `runMessageLoopOnPause()`? E.g. if a message was delivered indicating the caller wanted to invoke `console.time()` directly from the JS console?

Granted, _usually_ `IoContext::hasCurrent()` should be true there, but what if a breakpoint was set in global startup where there is no IoContext?

**Review by kentonv — APPROVED** [review]
I assume you've tested this running in the fiddle service (internal codebase)? That should be a pretty easy way to verify it works with the internal threading model, in addition to the workerd model.

**Comment by kentonv** [comment]
(Sorry for the slow review, this fell off my radar last week somehow.)

**Comment by paplco** [comment]
awesome, approved! 
hopefully this fixes setting breakpoints in vscode
is there a target release to include this PR?
Thanks!

**Comment by ohodson** [comment]
> I assume you've tested this running in the fiddle service (internal codebase)? That should be a pretty easy way to verify it works with the internal threading model, in addition to the workerd model.

Yes, you can set breakpoints and single step. Logging to console and heap querying appear to still work as before.

File: src/workerd/io/worker.c++ (hunk 18 above)
Comment (ohodson) [reply src/workerd/io/worker.c++:453]: I've not seen a case where we recurse into this.

Devtools proper on regular web pages and ew fiddle behave the same way: `console.time()` returns "undefined" when it is called from the console; the second time it is invoked it also logs a warning: `Timer 'default' already exists`.

There is a potential `ew-bin-test` to follow. It might be good to have a workerd test too. If we could launch isolates within a config in separate threads, then we could have a wd-test with two isolates: a worker to be tested and another that sends CDP messages to it to exercise functionality.

File: src/workerd/io/worker.c++ (hunk 17 above)
Comment (ohodson) [reply src/workerd/io/worker.c++]: Done, but PTAL as it's a first conversion. Thanks.

File: src/workerd/io/worker.c++ [hunk 19]
```diff
@@ -2316,8 +2300,237 @@ public:
...
+      }
+    }
+
+    kj::Promise<void> outgoingLoop() {
+      for (;;) {
+        co_await outgoingQueueNotifier->awaitNotification();
+        try {
+          auto lockedOutgoingQueue = outgoingQueue.lockExclusive();
+          if (lockedOutgoingQueue->status == MessageQueue::Status::CLOSED) {
+            co_await webSocket.close(1000, "client closed connection");
```
Comment (kentonv) [inline src/workerd/io/worker.c++:2467]: It looks like the behavior changed slightly in the coroutine conversion, previously we would send any remaining messages before closing.

File: src/workerd/io/worker.c++ [hunk 20]
```diff
@@ -2316,8 +2300,237 @@ public:
...
+          }
+        }
+      }
+    }
+
+    kj::Promise<void> outgoingLoop() {
+      for (;;) {
+        co_await outgoingQueueNotifier->awaitNotification();
+        try {
+          auto lockedOutgoingQueue = outgoingQueue.lockExclusive();
```
Comment (kentonv) [inline src/workerd/io/worker.c++:2465]: You need to drop this lock before co_awaiting.

File: src/workerd/io/worker.c++ (hunk 19 above)
Comment (ohodson) [reply src/workerd/io/worker.c++]: Ack, fixed, thanks.

File: src/workerd/io/worker.c++ (hunk 20 above)
Comment (ohodson) [reply src/workerd/io/worker.c++:2464]: Done.
